app/javascript/components/widget-transition-controls.jsx

```diff
@@ -0,0 +1,139 @@
+import React, { useEffect, useReducer } from 'react';
+import PropTypes from 'prop-types';
+import { mathMod } from 'ramda';
+
+const prefetchDelay = 1500;
+let noPrefetchRequestedYet = true;
+
+export const selectShouldRequestWidget = state =>
+  state.msRemainingSinceLastStart < Date.now() - state.startTimestamp;
+
+export const selectShouldRequestPrefetch = state =>
+  state.msRemainingSinceLastStart - (Date.now() - state.startTimestamp) <
+  prefetchDelay;
+
+export const selectIsPaused = state => state.startTimestamp === null;
+
+const reducer = (state, action) => {
+  const { startTimestamp, msRemainingSinceLastStart } = state;
+  const now = Date.now();
+
+  switch (action.type) {
+    case 'togglePause':
+      return {
+        msRemainingSinceLastStart: selectIsPaused(state)
+          ? msRemainingSinceLastStart
+          : msRemainingSinceLastStart - (now - startTimestamp),
+        startTimestamp: selectIsPaused(state) ? now : null,
+      };
+    case 'startTimer':
+      return {
+        msRemainingSinceLastStart: action.msDurationSeconds,
+        startTimestamp: selectIsPaused(state) ? null : now,
+      };
+    default:
+      return state;
+  }
+};
+
+const WidgetTransitionControls = ({
+  WrappedComponent,
+  requestPrefetch,
+  requestWidget,
+  cityName,
+  widgets,
+}) => {
+  const { byIdOrFirst: current, enabled: enabledWidgets } = widgets;
+
+  const widgetTransitionTime = current.durationSeconds * 1000;
+
+  const index = enabledWidgets.findIndex(w => w.id === current.id);
+
+  const prevWidgetId =
+    enabledWidgets[mathMod(index - 1, enabledWidgets.length)].id;
+  const nextWidgetId =
+    enabledWidgets[mathMod(index + 1, enabledWidgets.length)].id;
+
+  const [state, dispatch] = useReducer(reducer, {
+    startTimestamp: Date.now(),
+    msRemainingSinceLastStart: null,
+  });
+
+  useEffect(() => {
+    noPrefetchRequestedYet = true;
+    dispatch({
+      type: 'startTimer',
+      msDurationSeconds: widgetTransitionTime,
+    });
+  }, [current.id]);
+
+  useEffect(() => {
+    const timerId = setInterval(() => {
+      if (selectIsPaused(state)) {
+        return;
+      }
+      if (selectShouldRequestWidget(state)) {
+        requestWidget(nextWidgetId);
```

Comment:
maybe if we dispatch something like this:
```js
dispatch({type:'preload', func: requestWidget })
```

Then we could setup the action in our reducer like this:
```js
...
case 'preload':
  if(!state.preloaded) {
    action.func(selectNextWidgetId(state))
  }
  return {
    preloaded: true,
   ...state,
  }
...
```

Reply:
@chewnoill I implemented a dispatch option for 'prefetch', and also moved `nextWidgetId` and `prevWidgetId` computations into a resolver function.